rpcs3/Emu/Cell/lv2/sys_uart.cpp

```diff
@@ -1,35 +1,597 @@
+	}
+
+	if (size > 0x20000U)
+	{
+		// kmalloc restriction
+		fmt::throw_exception("Buffer is too big");
+	}
+
+	if (mode > 1)
+	{
+		return CELL_EINVAL;
+	}
+
+	const std::unique_ptr<u8[]> data = std::make_unique<u8[]>(size);
+	u32 read_size = 0;
+
+	auto vuart_read = [&](u8 *buf, u32 buf_size) -> s32
+	{
+		const u32 ITER_SIZE = 4096;
+		std::unique_lock lock(vuart_thread.rx_mutex, std::defer_lock);
+
+		if (!lock.try_lock())
+		{
+			return CELL_EBUSY;
+		}
+
+		const u32 first_iter_size = std::min(ITER_SIZE, buf_size);
+		const u32 first_iter_read = vuart_thread.read_rx_data(buf, first_iter_size);
+
+		if (first_iter_read < first_iter_size || buf_size <= ITER_SIZE)
+		{
+			return first_iter_read;
+		}
+
+		u32 read_size;
+
+		for (read_size = ITER_SIZE; read_size + ITER_SIZE < buf_size; read_size += ITER_SIZE)
+		{
+			const u32 in_loop_read = vuart_thread.read_rx_data(buf + read_size, ITER_SIZE);
+
+			if (in_loop_read < ITER_SIZE)
+			{
+				return read_size + in_loop_read;
+			}
+		}
+
+		return read_size + vuart_thread.read_rx_data(buf + read_size, buf_size - read_size);
+	};
+
+	if (mode == 1)
+	{
+		// Yield before checking for packets
+		lv2_obj::sleep(ppu);
+
+		for (;;)
+		{
+			if (ppu.is_stopped())
+			{
+				return {};
+			}
+
+			std::unique_lock<shared_mutex> lock(vuart_thread.rx_wake_m);
+			const s32 read_result = vuart_read(data.get(), static_cast<u32>(size));
+
+			if (read_result > CELL_OK)
+			{
+				read_size = read_result;
+				break;
+			}
+
+			vuart_thread.rx_wake_c.wait_unlock(5000, lock);
+		}
+
+		ppu.check_state();
+	}
+	else
+	{
+		const s32 read_result = vuart_read(data.get(), static_cast<u32>(size));
+
+		if (read_result <= CELL_OK)
+		{
+			return read_result;
+		}
+
+		read_size = read_result;
+	}
+
+	if (!buffer || !vm::check_addr(buffer.addr(), vm::page_writable, read_size))
+	{
+		return CELL_EFAULT;
+	}
+
+	memcpy(buffer.get_ptr(), data.get(), read_size);
+	return not_an_error(read_size);
 }
 
-error_code sys_uart_send(vm::cptr<void> buffer, u64 size, u64 flags)
+error_code sys_uart_send(ppu_thread &ppu, vm::cptr<void> buffer, u64 size, u32 mode)
 {
-	sys_uart.todo("sys_uart_send(buffer=0x%x, size=0x%llx, flags=0x%x)", buffer, size, flags);
+	sys_uart.trace("sys_uart_send(buffer=0x%x, size=0x%llx, mode=0x%x)", buffer, size, mode);
 
-	return CELL_OK;
+	if (!size)
+	{
+		return CELL_OK;
+	}
+
+	if ((mode & ~3ul) != 0)
+	{
+		return CELL_EINVAL;
+	}
+
+	auto &vuart_thread = g_fxo->get<vuart_av>();
+
+	if (!vuart_thread.init)
+	{
+		return CELL_ESRCH;
+	}
+
+	if (size > 0x20000U)
+	{
+		// kmalloc restriction
+		fmt::throw_exception("Buffer is too big");
+	}
+
+	if (!vm::check_addr(buffer.addr(), vm::page_readable, static_cast<u32>(size)))
+	{
+		return CELL_EFAULT;
+	}
+
+	const std::unique_ptr<u8[]> data = std::make_unique<u8[]>(size);
+	memcpy(data.get(), buffer.get_ptr(), size);
+
+	std::unique_lock lock(vuart_thread.tx_mutex, std::defer_lock);
+
+	const u32 ITER_SIZE = 4096;
+
+	if (mode & 1)
```

Comment:
By the way, if you have any idea what the mode bits do, you should create an enum and use them here. I can see operational differences between the modes, but no idea if its consistent between read and write, or what exactly the bits mean. Even if the official name is unknown, just make up something to ease reading.

Reply:
Done